### ISSUE TITLE
Bookmarklet support

### ISSUE THREAD

**Issue by myfonj** [JasonBarnabe/greasyfork#764]
Introduce possibility to publish script as a third and perhaps most simple, surely oldest mechanism to teach browser new tricks: `javascript:` links, [bookmarklets](https://en.wikipedia.org/wiki/Bookmarklet).

---

I think idea is clear to all already, so just some musings around:

### Motivation:

There is no "authoritative" platform for this yet (what is almost unbelievable) and Greasyfork appears to be perfect fit for it. (There are many more or less personal collections scattered around internet, what is great, but common registry for it would not spoil anything.)

### Obvious advantages:
- No extension needed, just drag and drop / bookmark link.
- For one-time user-initiated action invoking item from bookmarks toolbar seems to be more convenient than digging userscript action.

### Technical details:

- Presumably same restrictions as userscripts regarding loading external scripts.
- Publishing formatted, installing inlined, possibly with "metadata" comment pointing to GF page.
- (Good bookmarklet editor or even viewer would be a godsend.)
- BTW I assume most of existing userscripts would work as bookmarklets even without modification (?) but haven't thought it would be viable just to add this option everywhere.

### Ideas:

- "installed" bookmarklets list for registered users that would *remember version when added* and notify on update. I don't think bookmarklets are updated very often, but *some* manual update easing mechanism would be nice.

**Comment by eight04** [comment]
I dropped bookmarklets when CSP became a common thing:
https://bugzilla.mozilla.org/show_bug.cgi?id=866522

**Comment by myfonj** [comment]
Yes, that issue is famous problem, but as I understand it, it affect only those *"envelope"* bookmarklets that tries to **inject** some scripts into document. This operation is blocked on CSP protected pages even for userscripts, isn't it?

Plain and simple bookmarklets with no external dependencies never stopped to work, AFAIK. -- at least I am happily using them on Twitter, Google and anywhere. I suppose the only pages where it would not work are those with totally forbidden scripting, but I have not encountered any yet.

---

The only real modern nuisance regarding `javascript:` links is that you cannot run them in **URL bar** directly anymore, what is understandable self-XSS mitigation move, but FMPOV sucks that you cannot allow it like "I know what I am doing". But it doesn't mean they don't work: they never stopped, but now you really have to invoke the bookmark item.

**Comment by jixunmoe** [comment]
Downside of Bookmarklet I can think of:

* Hard to maintain
* Can't auto update
* `@require` won't work
* Scripts requires early execution time is impossible.
* The user need to click the scripts they want to run everytime he or she refresh or navigated away.

**Comment by myfonj** [comment]
Sure, those are all quite obvious traits of bookmarklets.  Good to have them expressed, thanks! Just to muse around them:

- _hard maintenance_
  - those are just code like any other code published through GF. It's the URL encoded representation that's scary, but that's the point of this issue, to be able to publish "formatted" JS that would end up in "inline" and "installable" URL. I'm thinking just making it inline, maybe with checked/forced ASI and converted inline comments to ensure it will work in one line, not uglifying into undecipherable mess. Also editor should ideally be able to format inlined code to more or less original code form. (If conversion did just URI encoding, line breaks preserved (as `%0A`) so no need to alter anything other that that, but it would seem a bit strange and I am not sure how browsers cope with it.)
- _no autoupdating_
  - sure, you cannot change anyone's bookmark remotely.  So like I suggested, you could have personal registry ("faves") that could tell you if some got an update since you bookmarked them. But I see it as "nice to have". I think URL of script page in bookmarklet URI (`javascript:/* name URI*/void ...`) would be enough to allow users to get back and "check for update" when bookmarklet stop working.
- _no `@require`_
  - Indeed, could be inlined "by compiler", but it is a terrible idea. Point of simple bookmarklets is not building extensions-like programs. Plus there is some murky (yet huge) URL size limit anyway, so it would be good to stick really to "few KiB max" scripts.
- _no autorun_
  - sure, but that's also point of this: for simple scripts you run either just occasionally / nearly never, or very often but always "on demand" after page fully loaded, bookmarklets (or custom buttons, RIP) are ideal.

So those do not seem like counterarguments to me at all.

**Comment by JasonBarnabe** [comment]
My overall thought here is that this is not something that will be presented to users normally due to all the drawbacks it has, but I would be willing to make changes that make it work for those who are interested.

Is there a length limit on bookmarklets? One point against the concept is that to generate a javascript: link containing the entire script, the entire script would need to be in the body of the page, which is a significant strain for the server and client. Could get around this with a "generate bookmarklet link" kind of thing.

**Comment by JasonBarnabe** [comment]
As mentioned above, open to making specific changes to support this for those who are interested, but this won't be built-in by default due to the numerous drawbacks of the format.
